src/Spec.ts

```diff
@@ -539,6 +548,240 @@ export default class Spec {
+      // so that we still have this information at that point
+      // rather than having to reconstruct it, approximately
+      // ... someday!
+      const warn = (message: string) => {
+        const ruleId = 'completion-invocation';
+
+        const path = [];
+        let pointer: HTMLElement | null = xref.node;
+        let alg;
+        while (pointer != null) {
+          if (pointer.tagName === 'LI') {
+            // @ts-ignore
+            path.unshift([].indexOf.call(pointer.parentElement!.children, pointer));
+          }
+          if (pointer.tagName === 'EMU-ALG') {
+            alg = pointer;
+            break;
+          }
+          pointer = pointer.parentElement;
+        }
+        if (alg == null || !{}.hasOwnProperty.call(alg, 'ecmarkdownTree')) {
+          let pointer: Element | null = xref.node;
+          while (this.locate(pointer) == null) {
+            pointer = pointer.parentElement;
+            if (pointer == null) {
+              break;
+            }
+          }
+          if (pointer == null) {
+            this.warn({
+              type: 'global',
+              ruleId,
+              message: message + ` but I wasn't able to find a source location to give you, sorry!`,
+            });
+          } else {
+            this.warn({
+              type: 'node',
+              node: pointer,
+              ruleId,
+              message,
+            });
+          }
+        } else {
+          // @ts-ignore
+          const tree = alg.ecmarkdownTree as ReturnType<typeof parseAlgorithm>;
+          let stepPointer: OrderedListItemNode | UnorderedListItemNode = {
+            sublist: tree.contents,
+          } as OrderedListItemNode;
+          for (const step of path) {
+            stepPointer = stepPointer.sublist!.contents[step];
+          }
+          this.warn({
+            type: 'contents',
+            node: alg,
+            ruleId: 'invocation-return-type',
+            message,
+            nodeRelativeLine: stepPointer.location.start.line,
+            nodeRelativeColumn: stepPointer.location.start.column,
+          });
+        }
+      };
+
+      const consumedAsCompletion = isConsumedAsCompletion(xref);
+      // checks elsewhere ensure that well-formed documents never have a union of completion and non-completion, so checking the first child suffices
+      // TODO: this is for 'a break completion or a throw completion', which is kind of a silly union; maybe address that in some other way?
+      const isCompletion =
+        returnType.kind === 'completion' ||
+        (returnType.kind === 'union' && returnType.types[0].kind === 'completion');
+      if (['Completion', 'ThrowCompletion', 'NormalCompletion'].includes(xref.aoid)) {
+        if (consumedAsCompletion) {
+          // TODO mf to improve message
+          warn(
+            `${xref.aoid} is explicitly a Completion Record and should not be unwrapped or marked`
```

Comment:
```suggestion
            `${xref.aoid} clearly creates a Completion Record; it does not need to be marked as such, and it would not be useful to immediately unwrap its result`
```